### PR TITLE
Added type definition for hexo

### PR DESCRIPTION
**Pull request by KentarouTakeda**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/hexo/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,976 @@
...
+            register(name: string, fn: (args: ParsedArgs, fn: (err: any) => void) => void): void;
+        }
+
+        interface Processor {
+            register(pattern: RegExp | string | ((str: string) => any), fn: (file: Box.File) => void): void;
+            register(fn: (file: Box.File) => void): void;
+        }
+
+        interface Renderer {
+            register(srcExt: string, outExt: string, fn: (data: HexoRendererData, options?: any) => string, sync: true): void;
```
Comment (PranavSenthilnathan) [inline types/hexo/index.d.ts]: As discussed [here](http://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html), it's not recommended to have optional parameters in callbacks.

File: types/hexo/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,976 @@
...
+import moment = require('moment');
+import { ParsedArgs } from 'minimist';
+import Logger = require('bunyan');
+import underscore = require('underscore');
+import connect = require('connect');
+import Stream = require('stream');
+import util = require('hexo-util');
+import fs = require('fs');
+
+declare global {
```
Comment (PranavSenthilnathan) [inline types/hexo/index.d.ts]: Does this need to be declared in the global scope?

**Review by PranavSenthilnathan — CHANGES_REQUESTED** [review]

File: types/hexo/index.d.ts (hunk 2 above)
Comment (KentarouTakeda) [reply types/hexo/index.d.ts]: Thank you for your review. I think it is necessary.

Hexo provides a plugin system. Plugin works inside hexo. Then global value `hexo` is provided by hexo.
https://hexo.io/docs/themes#scripts

* TypeScript  
  https://github.com/OpportunityLiu/hexo-renderer-ts/blob/master/index.ts
* JavaScript  
  https://github.com/hexojs/hexo-deployer-git/blob/master/index.js

If there is any other good way, please let me know. I fill follow it.

File: types/hexo/index.d.ts (hunk 1 above)
Comment (KentarouTakeda) [reply types/hexo/index.d.ts]: Thank you!I will fix it.

Comment (KentarouTakeda) [reply types/hexo/index.d.ts]: I fixed it. Thank you!

**Review by PranavSenthilnathan — APPROVED** [review]

**Comment by KentarouTakeda** [comment]
Thank you for the review and merge!

**Pull request merged**